### PR TITLE
Twir add-on

### PR DESCRIPTION
**Pull request by crashmax-dev**
This add-on for https://twir.app adds a command suggestion.
image

### CODE REVIEW

File: src/twir/index.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+				description,
+				// parse `command.permisions`
+				permissionLevel: 0,
+				ffz_group: `Twir (${command.module})`,
+			}
+		})
+	}
+
+	unregisterRoomCommands() {
+		this.tabCommands = [];
```
Comment (SirStendec) [inline src/twir/index.js:75]: So, when a `chat:room-remove` event is being fired, you're just clearing your list of commands. This is, in theory, fine.

The problem is that when a user navigates to a new channel, `chat:room-add` is called immediately for the new room. `chat:room-remove` is called after a short delay. (There's a reference counting system in place for room data and removals happen after a short delay to protect against any quick leave-join-leave-join nonsense.)

In practice, this means that your tabCommands array is likely to be cleared when a user navigates between channels unless your API is particularly slow. Maybe store the commands in a map by room ID, and flatten it into an array as necessary? This makes me notice that the get-tab-commands event doesn't include the current channel, so I'll try to add something for that next time I'm paying attention.

File: src/twir/index.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	}
+
+	preSendMessage(event) {
+		const message = event.message.trim();
+		if (message.startsWith('!')) return;
+
+		if (message.startsWith('/') && message.includes(ZWE_SYMBOL)) {
```
Comment (SirStendec) [inline src/twir/index.js:82]: I'd prefer if you were more specific with the test for `ZWE_SYMBOL` here. It should always be at the start such that a command is `/${ZWE_SYMBOL}something` so you could have a regular expression like
```javascript
// top of file
const COMMAND_MATCHER = new RegExp(`^/${ZWE_SYMBOL}(.+)`);

// later
const match = COMMAND_MATCHER.exec(message);
if ( match )
    event.message = `!${match[1]}`;
```

This protects against the case where a user happens to have your special character in their command for whatever reason, which could in theory happen with some commands (an example being a custom ban/timeout reason). Users can be weird so I try to anticipate stuff like that.

File: src/twir/manifest.json [hunk 3]
```diff
@@ -0,0 +1,13 @@
+{
+  "enabled": false,
```
Comment (SirStendec) [inline src/twir/manifest.json:2]: You should set `enabled` to `true` when submitting a new add-on, for it to be built. If you don't want the add-on to be included in the add-on's list you can set `unlisted` to `true` as well.

**Review by SirStendec — CHANGES_REQUESTED** [review]
Hey there! Thanks for submitting a new add-on.

The big thing here is that you need to set `enabled` to `true` in your manifest for the add-on to get built. Everything else is technically fine, but I suggest you change how you're handling when rooms are added/removed as discussed in my comments.

As an aside, I think this is super neat. I've always intended to eventually have a bot command tab-completion thing in FrankerFaceZ but I've never actually gotten around to it. I might have to see if I can mess with the command autocomplete provider to let us just specify a command uses something other than `/` in order to remove the need for a pre-send event here.

**Comment by SirStendec** [comment]
Update: I added support for ! commands in 8807e09ea324db3def925a1e79aa05865879dd6b

This should simplify your logic a lot, since you can just specify that your commands have `prefix: '!'` and they'll be displayed/inserted/completed that way.

Example:
```javascript
ffz.on('chat:get-tab-commands', e => e.commands.push({
    prefix: '!',
    name: '8ball',
    description: 'oooooo spoookey',
    permissionLevel: 0,
    ffz_group: 'Quoteconut (Bot)',
    commandArgs: [
        {name: 'question', isRequired: true}
    ]
}));
```
image

File: src/twir/index.js (hunk 2 above)
Comment (crashmax-dev) [reply src/twir/index.js]: `ZWE_SYMBOL` outdated

**Pull request merged**